### PR TITLE
Ahoy containers

### LINKED ISSUE #8: Setting-up containerization
## Issue

As of now, there's no way to run projects in a container environment. This is critical to enable proper evaluation and project publication.

## Proposed solution

Implement `docker-compose` files to enable local and development environments deployments.

### PR DESCRIPTION
**Pull request by ruiconti**
This PR (partly) resolves #8.

## What this adds

Mainly, it implements a few things:

1. Local-development service execution can now be fully abstracted w/ `docker-compose`.
2. Local-testing can now be done without the need to deal with python environments
2. Makefile has been updated to help tooling.

### Memory footprint

Using debian's busters distro, a few checks that might be useful when we're benchmarking this app.

1. Idle-mode, meaning handling 0 requests, 8 workers:

image

2. Idle-mode, meaning handling 0 requests, single-worker:

image

Which is pretty straightforward, ~44MiB per worker of memory footprint. IMO it's definitely something work investigating i.e. what is taking so much space since this is supposed to be a lean framework.

### What is currently missing

For this PR, it is missing a few things that are being added as work is progressed:

- [x] Proper volume mapping
- [ ] Proper logging
- [x] Usage of poetry for dep mgmt

## What needs to be done

This work enables the service to be fully usable in local and development environments. So what is "missing" from this work in order to be used in production environments:

1. Security concerns (TLS everywhere)
2. Real infrastructure concerns (reversed proxies, load balancers)
3. CI/CD pipelines
4. Integration to cloud providers & image repositories
5. Proper & standardized logging

But those concerns would be addressed after we went public, IMO. But are kept in record so we have a clear vision of next steps.

@rafamelos @andreyrcdias care to review & complement?

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+version: "3.9"
+
+x-common-variables: &common-variables
+  POSTGRES_USER: ${POSTGRES_USER:-postgres}
+  POSTGRES_PASSWORD: ${POSTGRES_PASSWORD:-postgres}
+  POSTGRES_DB: ${POSTGRES_DB:-template}
+
+services:
+  dbpg: 
+    container_name: postgres
```
Comment (andreyrcdias) [inline docker-compose.yml:10]: I think we could have the prefix `kingdom_` or `"{PROJET_NAME}"` for each name of the containers, as u used in the `web` service

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    image: postgres
+    environment:
+      <<: *common-variables
+      PGDATA: /data/postgres
+    ports:
+      - "5432:5432"
+    networks:
+      - backend 
+    restart: unless-stopped
```
Comment (andreyrcdias) [inline docker-compose.yml:19]:
```suggestion
    restart: unless-stopped
    
    volumes:
      - kingdom-dbpg-data:/var/lib/postgresql/data/pgdata
```

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    ports:
+      - "5000:5000"
+    depends_on:
+      - dbpg 
+    volumes:
+      - ./:/app
+    networks:
+      - backend 
+    restart: unless-stopped
+
```
Comment (andreyrcdias) [inline docker-compose.yml:43]:
```suggestion
volumes:
  kingdom-dbpg-data:
```

**Review by andreyrcdias — COMMENTED** [review]
I commented on some suggestions about the names of the services.

About the volumes, I made some suggestions so that we can at that first moment, persist the data somewhere.

In the case of high memory usage, correct me if I'm wrong, is this related in any way to all the `dev` dependencies that are installed?

**Review by rafamelos — COMMENTED** [review]
I don't have much knowledge about using Docker (I definitely need some remembering and new studies on it), with that said I think I can't really give a appropriated opinion, but this seems good to me.  
About the memory footprint, do you have some refs our thoughts on what level of memory use we should have? I think a good approach to investigate this would be to strip the project removing everything and running only with starlette, and again only with starlette and ariadne, in a way that we can learn what is the minimum consumption we can achieve with our current choices.

File: docker-compose.yml (hunk 1 above)
Comment (ruiconti) [reply docker-compose.yml:10]: Agreed!

File: docker-compose.yml (hunk 2 above)
Comment (ruiconti) [reply docker-compose.yml:19]: Do we need to create this directory locally?

Comment (andreyrcdias) [reply docker-compose.yml:19]: No, docker creates this for us

**Comment by ruiconti** [comment]
> I commented on some suggestions about the names of the services.
> About the volumes, I made some suggestions so that we can at that first moment, persist the data somewhere.

Thanks! Those were included.

> In the case of high memory usage, correct me if I'm wrong, is this related in any way to all the dev dependencies that are installed?

No way to tell without investigating. Created #16 to track that.

> About the proper logging and poetry dependency management, I need to do a study and see the best practices

Gonna leave that for future improvements as they aren't critical.

**Pull request merged**